### PR TITLE
Fix performance regression in task plan execution

### PR DESCRIPTION
**Pull request by wolfs**

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java [hunk 1]
```diff
@@ -125,12 +125,13 @@ boolean hasUnlockedResources() {
 
         @Override
         public void releaseLocks() {
-            if (lockedResources != null) {
+            if (lockedResources != null && !lockedResources.isEmpty()) {
```
Comment (oehme) [inline subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java]: This is not really necessary - the for loop will just short-circuit anyway.

File: subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java [hunk 2]
```diff
@@ -216,7 +217,7 @@ public boolean hasTask(String path) {
 
     public List<Task> getAllTasks() {
         ensurePopulated();
-        return taskExecutionPlan.getTasks();
+        return ImmutableList.copyOf(taskExecutionPlan.getTasks());
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: I'd cache this once it's been used once.

File: subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java (hunk 1 above)
Comment (wolfs) [reply subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java]: I removed it again.

File: subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: Caching now.

File: subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java [hunk 3]
```diff
@@ -216,7 +218,10 @@ public boolean hasTask(String path) {
 
     public List<Task> getAllTasks() {
         ensurePopulated();
-        return taskExecutionPlan.getTasks();
+        if (allTasks == null) {
+            allTasks = ImmutableList.copyOf(taskExecutionPlan.getTasks());
+        }
+        return allTasks;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: I don't think caching this is a good idea. The plan can get recalculated after `addTasks()` is called. We also don't call this method a lot, so it's probably not worth caching in the first place.

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionGraph.java]: Good point. I fixed the problem by resetting the cache in `ensurePopulated()`.

**Pull request merged**